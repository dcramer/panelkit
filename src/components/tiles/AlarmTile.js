import PropTypes from "prop-types";

import Tile, { TileProps } from "../Tile";

export default class AlarmTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getIcon() {
    if (this.props.icon) return this.props.icon;
    const { state } = this.getEntity(this.props.entityId);
    return state === "armed_home" || state === "armed_away"
      ? "bell"
      : "bell-off";
  }

  getWatchedEntityIds() {
    return [this.props.entityId];
  }
}
