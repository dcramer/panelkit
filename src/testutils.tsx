import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
