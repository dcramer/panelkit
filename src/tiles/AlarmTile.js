import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";

export default class AlarmTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    return state === "armed_home" || state === "armed_away"
      ? "bell"
      : "bell-off";
  }
}
