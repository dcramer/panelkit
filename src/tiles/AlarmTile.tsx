import React from "react";

import Tile, { TileProps } from "./Tile";
import AlarmModal, { AlarmModalProps } from "../modals/AlarmModal";

const ARMED_STATES = new Set(["armed_home", "armed_away", "armed_night"]);

type AlarmTileProps = TileProps & {
  entityId: string;
};

export default class AlarmTile extends Tile<AlarmTileProps> {
  onTouch = async () => {
    const { state } = this.getEntity(this.props.entityId);
    if (ARMED_STATES.has(state)) {
      await this.callService(
        "alarm_control_panel",
        "alarm_disarm",
        {
          entity_id: this.props.entityId,
        },
        {
          [this.props.entityId]: {
            state: "disarmed",
          },
        }
      );
    } else {
      this.openModal();
    }
  };

  renderModal(params: AlarmModalProps) {
    return <AlarmModal {...params} />;
  }

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    switch (state) {
      case "armed_home":
      case "armed_away":
      case "armed_night":
        return "bell";
      case "triggered":
        return "bell-ring";
      case "disarmed":
        return "bell-off";
      default:
        return "bell";
    }
  }
}
