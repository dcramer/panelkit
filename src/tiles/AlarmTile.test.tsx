import { waitFor } from "@testing-library/react";

import { renderTile, simulateTouch } from "../testutils";
import AlarmTile from "./AlarmTile";

test("renders", async () => {
  const { container, getByText } = renderTile(AlarmTile, {
    entity_id: "alarm_control_panel.test",
    state: "disarmed",
    attributes: {
      friendly_name: "Test Alarm",
    },
  });

  expect(getByText(/Test Alarm/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("disarms on touch event", async () => {
  const { container, hass } = renderTile(AlarmTile, {
    entity_id: "alarm_control_panel.test",
    state: "armed_home",
    attributes: {},
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "alarm_control_panel",
      "alarm_disarm",
      {
        entity_id: "alarm_control_panel.test",
      },
      {
        "alarm_control_panel.test": { state: "disarmed" },
      }
    );
  });
});
