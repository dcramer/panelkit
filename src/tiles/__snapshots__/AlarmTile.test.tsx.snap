// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-disarmed touchable"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Disarmed
    </div>
    <div
      class="tile-body"
    >
      <span
        class="tile-icon"
      >
        <svg />
      </span>
    </div>
    <div
      class="tile-title"
    >
      Test Alarm
    </div>
  </div>
</div>
`;
